src/Compilers/Shared/BuildServerConnection.cs

```diff
@@ -595,4 +612,216 @@ public static string GetTempPath(string workingDir)
         }
     }
+
+    internal interface IServerMutex : IDisposable
+    {
+        bool TryLock(int timeoutMs);
+        void Unlock();
+        bool IsDisposed { get; }
+    }
+
+    internal sealed class FileMutex : IDisposable
+    {
+        public readonly bool OwnsMutex;
+        public readonly FileStream Stream;
+        public readonly string FilePath;
+
+        public bool IsLocked { get; private set; }
+
+        internal static string GetMutexDirectory()
+        {
+            var homeDirectory = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);
+            if (String.IsNullOrWhiteSpace(homeDirectory))
+                homeDirectory = Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData);
+            if (String.IsNullOrWhiteSpace(homeDirectory))
+                homeDirectory = Path.GetTempPath();
+            var result = Path.Combine(homeDirectory, ".roslyn");
+            if (!Directory.Exists(result))
+                Directory.CreateDirectory(result);
+            return result;
+        }
+
+        public FileMutex(string name)
+        {
+            FilePath = Path.Combine(GetMutexDirectory(), name);
+            try
+            {
+                Stream = new FileStream(FilePath, FileMode.Open, FileAccess.ReadWrite, FileShare.None);
+                OwnsMutex = false;
+            }
+            catch (FileNotFoundException)
+            {
+                Stream = new FileStream(FilePath, FileMode.CreateNew, FileAccess.ReadWrite, FileShare.None);
+                OwnsMutex = true;
+            }
+        }
+
+        public bool TryLock(int timeoutMs)
+        {
+            if (IsLocked)
+                throw new InvalidOperationException("Lock already held");
+
+            var sw = Stopwatch.StartNew();
+            do
+            {
+                try
+                {
+                    Stream.Lock(0, 0);
+                    IsLocked = true;
+                    return true;
+                }
+                catch (IOException)
+                {
+                    // Lock currently held by someone else
+                    Thread.Sleep(0);
+                }
+                catch (Exception)
+                {
+                    // Something else went wrong.
+                    return false;
+                }
+            } while (sw.ElapsedMilliseconds < timeoutMs);
+
+            return false;
+        }
+
+        public void Unlock()
+        {
+            if (!IsLocked)
+                return;
+            Stream.Unlock(0, 0);
+            IsLocked = false;
+        }
+
+        public void Dispose()
+        {
+            var wasLocked = IsLocked;
+            if (wasLocked)
+                Unlock();
+            Stream.Dispose();
+            // We do not delete the lock file here because there is no reliable way to perform a
+            //  'delete if no one has the file open' operation atomically on *nix. This is a leak.
```

Comment:
Can you elaborate more here about why no delete can occur? What goes wrong if the file is deleted right now?

Reply:
If you delete it when someone has it open, you're basically unlocking the mutex for them, and the next person who comes along won't see a mutex at all so they'll create a new one and lock it.

Reply:
The delete should be blocked though as you're using `FileShare.None`. Any attempt to delete while the other process has the file open with `FileShare.None` should result in an ecxetpion, not a silent delete. 

---
In reply to: [253236703](https://github.com/dotnet/roslyn/pull/31497#discussion_r253236703) [](ancestors = 253236703)

Reply:
I thought so, but that's not what happens. Maybe my code had a bug.

Reply:
OK, I went and checked again because I recalled that this is expected for Linux:
https://bugzilla.xamarin.com/show_bug.cgi?id=38201
There is no posix primitive for us to use here, I think? Same reason abandoned mutex detection isn't possible without coreclr's extremely complex workaround. I'm not aware of any way to stop someone from rm'ing a file I have open. The only thing I could come up with is grepping lsof to find anyone else with the lock open, but the performance implications of that are profound and it's gross (even though the documentation suggests that lsof is meant to be used this way)

.Lock(0, 0) and .Unlock work everywhere, which is why I'm using them, but sadly I can't use those to block deletion - you can still delete a file that has a lock held on it, and then create a new file with the same name in its place.

Reply:
Do you need me to find some sort of way to emulate the FileShare.None constraint so we can automatically delete the lock file on server shutdown? I'm not really sure how it would be done, so if it's a requirement I'll have to do more research. Would we be willing to tolerate shelling out to lsof, calling native posix APIs or opening files in ```/proc/``` to do it?

Reply:
> Do you need me to find some sort of way to emulate the FileShare.None constraint s

No. I chatted with some others about this to better understand the .NET Core behavior here and see if there was a pattern we could copy into Mono. They had the exact same feedback you did: there is just no reasonable way to emulate this on Linux / Mac. 

I'm still a bit bothered by the leak though. One idea I had is this: what if the server process deleted them all on shutdown? The server process should always be outliving the client processes here. Is there a big hole that gets opened up if the server deletes them during shut down?

Reply:
I think there's a slight risk of an arcane race there where a server that's shutting down deletes the lock file created a server that just started up, but that's all I can think of. Otherwise, we can do that and this just stops being a generalized file mutex (which we can probably just make clear by changing naming and/or adding comments)

Reply:
Do you want me to change it so the server deletes the lock on shutdown?

Reply:
Let's just file an item to track that.